File: beatrice.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Beatrice and the Dream Tree
--------------------------------------------
We must count how many nodes are uniquely identifiable
in a rooted tree (root = 1). Two nodes are identical
if a rooted-tree automorphism swaps them.

Approach (O(N)):
1. Compute subtree hash (down-hash) bottom-up.
2. Compute "up-hash" (parent-side hash) top-down.
3. Each node’s full signature = multiset of its children’s
   hashes + parent branch (if exists).
4. Count how many signatures are unique.

We use 64-bit SplitMix64 hashing for low collision risk.
*/

using u64 = unsigned long long;
const u64 MASK = (1ULL << 64) - 1;

// --- SplitMix64 deterministic hash ---
inline u64 splitmix64(u64 x) {
    x += 0x9e3779b97f4a7c15ULL;
    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9ULL;
    x = (x ^ (x >> 27)) * 0x94d049bb133111ebULL;
    x ^= (x >> 31);
    return x;
}
inline u64 rnd(u64 x) { return splitmix64(x + 0x9e3779b97f4a7c15ULL); }
inline u64 mix(u64 sum_s, int cnt) { return splitmix64(sum_s ^ splitmix64(cnt)); }

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    if (!(cin >> N)) return 0;

    vector<vector<int>> g(N + 1);
    for (int i = 0; i < N - 1; ++i) {
        int u, v;
        cin >> u >> v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    vector<int> parent(N + 1, 0);
    vector<int> order; order.reserve(N);
    parent[1] = -1;
    queue<int> q; q.push(1);

    while (!q.empty()) {
        int u = q.front(); q.pop();
        order.push_back(u);
        for (int w : g[u]) if (!parent[w]) {
            parent[w] = u;
            q.push(w);
        }
    }

    vector<vector<int>> children(N + 1);
    for (int i = 2; i <= N; ++i)
        children[parent[i]].push_back(i);

    vector<u64> down(N + 1), up(N + 1), sum_children(N + 1);
    vector<int> cnt_children(N + 1);
    vector<char> has_up(N + 1, 0);

    // ---- Bottom-up pass ----
    for (int i = N - 1; i >= 0; --i) {
        int u = order[i];
        u64 s = 0; int c = 0;
        for (int v : children[u]) {
            s += rnd(down[v]);
            ++c;
        }
        down[u] = mix(s, c);
        sum_children[u] = s;
        cnt_children[u] = c;
    }

    // ---- Top-down pass ----
    for (int u : order) {
        for (int v : children[u]) {
            u64 s_excl = (sum_children[u] - rnd(down[v]));
            int c_excl = cnt_children[u] - 1;
            if (has_up[u]) {
                s_excl += rnd(up[u]);
                ++c_excl;
            }
            up[v] = mix(s_excl, c_excl);
            has_up[v] = 1;
        }
    }

    // ---- Compute full signatures ----
    unordered_map<u64, int> freq;
    freq.reserve(N * 2);
    vector<u64> full_sig(N + 1);

    for (int u = 1; u <= N; ++u) {
        u64 s = sum_children[u];
        int c = cnt_children[u];
        if (has_up[u]) {
            s += rnd(up[u]);
            ++c;
        }
        u64 sig = mix(s, c);
        full_sig[u] = sig;
        ++freq[sig];
    }

    // ---- Count unique signatures ----
    int ans = 0;
    for (int u = 1; u <= N; ++u)
        if (freq[full_sig[u]] == 1) ++ans;

    cout << ans << "\n";
    return 0;
}

File: beatrice.py
# girl should find how many nodes ..? (in respect to where she lands)
# Two nodes are identical if a tree automorphism swaps them.(ignore)
# for me - i sent this to nicole (2025/10/25)
# Solution - compute subtree hashes.


import sys
sys.setrecursionlimit(1 << 25)
rd = sys.stdin.buffer.readline

# -------- 64-bit deterministic hashing helpers --------
#chatgpt this part (much quicker than my original implementation)
# From https://en.wikipedia.org/wiki/SplitMix64

#going forward also put this in every code that needs hashing..

MASK = (1 << 64) - 1

def splitmix64(x: int) -> int:
    x = (x + 0x9e3779b97f4a7c15) & MASK
    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9 & MASK
    x = (x ^ (x >> 27)) * 0x94d049bb133111eb & MASK
    x ^= (x >> 31)
    return x & MASK

def rnd(x: int) -> int:
    # Deterministically "randomize" any 64-bit value.
    return splitmix64(x + 0x9e3779b97f4a7c15)

def mix(sum_s: int, cnt: int) -> int:
    # Combine a multiset's randomized-sum and its size into a single 64-bit hash.
    # Using splitmix on (sum_s XOR splitmix(cnt)) gives very low collision chance.
    return splitmix64((sum_s ^ splitmix64(cnt)) & MASK)

# ------------------------------------------------------

def main():
    N_line = rd()
    while N_line.strip() == b'':
        N_line = rd()
    N = int(N_line)

    # building undirected but connected list 
    g = [[] for _ in range(N + 1)]
    for _ in range(N - 1):
        u, v = map(int, rd().split())
        g[u].append(v)
        g[v].append(u)

    # Root the tree at 1 (the original tree is rooted at node 1)
    parent = [0] * (N + 1)
    parent[1] = -1
    order = [1]  # BFS/DFS order from root, used to avoid recursion

    for u in order:
        for w in g[u]:
            if w == parent[u]:
                continue
            parent[w] = u
            order.append(w)

    # Build children lists (directed by root)
    children = [[] for _ in range(N + 1)]
    for v in range(2, N + 1):
        children[parent[v]].append(v)

    # next step 
    # down[u] = hash of the multiset of children's down-hashes at u
    down = [0] * (N + 1)
    # For later reuse, also keep (sum over rnd(down[child]), child_count)
    sum_children = [0] * (N + 1)
    cnt_children = [0] * (N + 1)

    for u in reversed(order):
        s = 0
        for v in children[u]:
            s = (s + rnd(down[v])) & MASK
        c = len(children[u])
        down[u] = mix(s, c)
        sum_children[u] = s
        cnt_children[u] = c

    # next step 
    up = [0] * (N + 1)
    has_up = [False] * (N + 1)  # root has no parent branch

    for u in order:
        # For each child v of u, exclude v's own branch from u and include u's up-branch
        for v in children[u]:
            # Remove v's contribution from u's children multiset
            s_excl = (sum_children[u] - rnd(down[v])) & MASK
            c_excl = cnt_children[u] - 1
            # If u has a parent branch, include it as one more "child"
            if has_up[u]:
                s_excl = (s_excl + rnd(up[u])) & MASK
                c_excl += 1
            # Now the parent branch seen by v is simply the hash of this virtual node
            up[v] = mix(s_excl, c_excl)
            has_up[v] = True

    # ---------- next step ----------
    full_sig = [0] * (N + 1)
    freq = {}

    for u in range(1, N + 1):
        s = sum_children[u]
        c = cnt_children[u]
        if has_up[u]:
            s = (s + rnd(up[u])) & MASK
            c += 1
        sig = mix(s, c)
        full_sig[u] = sig
        freq[sig] = freq.get(sig, 0) + 1

    # Count uniquely identifiable nodes (those with a unique signature)
    answer = sum(1 for u in range(1, N + 1) if freq[full_sig[u]] == 1)
    print(answer)

if __name__ == "__main__":
    main()
